src/i_oalsound.c

```diff
@@ -0,0 +1,662 @@
+    }
+
+    for (i = 0; i < MAX_CHANNELS; i++)
+    {
+        alSourcei(sources[i], AL_SOURCE_RESAMPLER_SOFT, def_resampler);
+    }
+/*
+    printf(" Using '%s' resampler.\n",
+           alGetStringiSOFT(AL_RESAMPLER_NAME_SOFT, def_resampler));
+*/
+}
+
+void I_OAL_ResetSource2D(int channel)
+{
+    if (!oal)
+    {
+        return;
+    }
+
+    // Don't reset while sound is playing.
+    if (oal->active[channel])
+    {
+        return;
+    }
+
+    if (oal->SOFT_source_spatialize)
+    {
+        alSourcei(oal->sources[channel], AL_SOURCE_SPATIALIZE_SOFT, AL_FALSE);
+    }
+
+    if (oal->EXT_EFX)
+    {
+        alSourcef(oal->sources[channel], AL_AIR_ABSORPTION_FACTOR, 0.0f);
+    }
+
+    if (oal->EXT_SOURCE_RADIUS)
+    {
+        alSourcef(oal->sources[channel], AL_SOURCE_RADIUS, 0.0f);
+    }
+
+    alSource3f(oal->sources[channel], AL_POSITION, 0.0f, 0.0f, 0.0f);
+    alSource3f(oal->sources[channel], AL_VELOCITY, 0.0f, 0.0f, 0.0f);
+    alSource3i(oal->sources[channel], AL_DIRECTION, 0, 0, 0);
+
+    alSourcei(oal->sources[channel], AL_ROLLOFF_FACTOR, 0);
+    alSourcei(oal->sources[channel], AL_SOURCE_RELATIVE, AL_TRUE);
+}
+
+void I_OAL_ResetSource3D(int channel, boolean point_source)
+{
+    if (!oal)
+    {
+        return;
+    }
+
+    // Don't reset while sound is playing.
+    if (oal->active[channel])
+    {
+        return;
+    }
+
+    if (oal->SOFT_source_spatialize)
+    {
+        alSourcei(oal->sources[channel], AL_SOURCE_SPATIALIZE_SOFT, AL_TRUE);
+    }
+
+    if (oal->EXT_EFX)
+    {
+        alSourcef(oal->sources[channel], AL_AIR_ABSORPTION_FACTOR, oal->absorption);
+    }
+
+    if (oal->EXT_SOURCE_RADIUS)
+    {
+        alSourcef(oal->sources[channel], AL_SOURCE_RADIUS,
+                  point_source ? 0.0f : OAL_SOURCE_RADIUS);
+    }
+
+    alSourcei(oal->sources[channel], AL_ROLLOFF_FACTOR, OAL_ROLLOFF_FACTOR);
+    alSourcei(oal->sources[channel], AL_SOURCE_RELATIVE, AL_FALSE);
+}
+
+void I_OAL_AdjustSource3D(int channel, const ALfloat *position,
+                          const ALfloat *velocity, const ALfloat *direction)
+{
+    if (!oal)
+    {
+        return;
+    }
+
+    alSourcefv(oal->sources[channel], AL_POSITION, position);
+    alSourcefv(oal->sources[channel], AL_VELOCITY, velocity);
+    alSourcefv(oal->sources[channel], AL_DIRECTION, direction);
+}
+
+void I_OAL_AdjustListener3D(const ALfloat *position, const ALfloat *velocity,
+                            const ALfloat *orientation)
+{
+    if (!oal)
+    {
+        return;
+    }
+
+    alListenerfv(AL_POSITION, position);
+    alListenerfv(AL_VELOCITY, velocity);
+    alListenerfv(AL_ORIENTATION, orientation);
+}
+
+void I_OAL_UpdateUserSoundSettings(void)
+{
+    if (!oal)
+    {
+        return;
+    }
+
+    // Source parameters.
+    SetResampler(oal->sources);
+    oal->absorption = (ALfloat)snd_absorption / 2.0f;
+
+    // Context state parameters.
+    alDopplerFactor((ALfloat)snd_doppler * snd_doppler / 100.0f);
```

Comment:
The docs say that the default value is `1`. Does this mean that we can only deemphasize the Doppler effect by chosing a value less than `10`? And why the quadratic curve?

Reply:
At `1` the effect is already excessive and it's terrible at higher values than that. The quadratic curve is arbitrary: small values seem ideal but I also wanted to include the "default" value of `1`.

Reply:
You'll notice right above that is sound absorption. It's tuned to a 0dB, ~3dB, ~6dB drop off (thermo positions 0, 5, 10) when reaching the distance where the sound fades out. That seems like a reasonable limit.

Reply:
Indeed.

How a bout a split approach for the Doppler factor, which allows for fine tuning up to the already excessive value of `1` (i.e. `1/100`) at thermo position 5 and escalates quickly to to `100` after that.

For example:

```
(snd_doppler <= 5) ? (snd_doppler / 5) : exp(0.2 * (snd_doppler - 5) * ln(100.))
```

Reply:
Or, if `alDopplerFactor(1.0)` is really completely unreasonable, let's just scale linearly from e.g. `0` over `1./100.` to `2./100.` at thermo positions 0, 5, 10.

Reply:
Which value was used in the video you sent me? This should be close to thermo position 10, IMHO.

Reply:
> Which value was used in the video you sent me? This should be close to thermo position 10, IMHO.

It was 1.0, so yes, it seems like a good upper limit.

Reply:
I think I'm confused here, so maybe try the current thermo for the Doppler effect (quadratic) and then we can decide if it needs adjustment.

Reply:
I am also confused. 😆 with `1` do you mean `1.0` or `1.0/100.0`?

Reply:
The current thermo maxed out means `alDopplerFactor(1.0)` or `100.0/100.0`.

Reply:
Ah, okay, then it should work! I'll compile the code tomorrow and experiment with it.

Reply:
I just played a bit with the Doppler effect thermo set to 5 and the effect is subtle and nice.